framework/include/base/MooseApp.h

```diff
@@ -719,18 +719,45 @@ class MooseApp : public ConsoleStreamInterface,
    */
   const MooseMesh * masterDisplacedMesh() const { return _master_displaced_mesh; }
 
+  /**
+   * Class that is used as a parameter to setFinalMeshGeneratorName()
+   *  that allows only MeshGeneratorMesh methods to call this methods
+   */
+  class SetFinalMeshGeneratorNameKey
+  {
+    friend class MeshGeneratorMesh;
+    SetFinalMeshGeneratorNameKey() {}
+    SetFinalMeshGeneratorNameKey(const SetFinalMeshGeneratorNameKey &) {}
+  };
```

Comment:
complexity? six lines? clear intent? descriptive name? I disagree

Reply:
It's just not worth it. And the intent is hostile to the next developer. This conveys the intent that this code is only for the purposes here in this PR, and no one else.

Reply:
> This conveys the intent that this code is only for the purposes here in this PR, and no one else.

That's the exact purpose - this method should only be used for one specific case?

Reply:
Oh... and a static compilation error is always better than an assertion or a dynamic error. No need to create a test for it, no need to run a test to see if it works.

Reply:
> and a static compilation error is always better than an assertion or a dynamic error. 

I don't know if I would say that. A compilation error can often be pretty opaque compared to some dynamic error messages. For instance the private constructor of un-specialized `StandardType` does not give a developer much clue about what they might have done wrong (in most cases it is typically a forgotten include).

But that being said I feel like this code is pretty easy to understand. This is more fine-grained control than my first thought, which was making this method private, making `MeshGeneratorMesh` a friend of `MooseApp`, and allowing `MeshGeneratorMesh` access to all private data of `MooseApp`.

I don't look at this code and think "wow, that was hostile". I look at it and think it's pretty clever. I like it

Reply:
The MooseApp has "what am I doing now" routines that can be used to limit when each routine is called if you want some fine-grained control. 

You end up creating structs during your simulation (while the code is running) that are meant for building/compiling the code. It's totally unexpected (and I dont care that it's cheap or efficient, it's not the point)

Reply:
The structs are optimized out

Reply:
proof: https://godbolt.org/z/95c3ro65j

Reply:
This pattern is ubiquitous enough that Chromium has a page dedicated to it: https://chromium.googlesource.com/chromium/src/+/HEAD/docs/patterns/passkey.md

Reply:
The pattern is surprising enough that Chromium, a project that isnt really for novice developers to use APIs from, felt the need to add a page for it.

Reply:
I think we need to make a clear distinction to where these API restrictions are set.

These APIs are deep to execution handling in MOOSE. Recall that the purpose of MOOSE is to (1) make non computer scientists be able to create material models, kernels, meshes, etc, without a strong development background, (2) to enable multiphysics by providing a common framework between applications, and hopefully (3) to make a better place for strong developers to develop the internals that should not need to be touched by many.

How many of our target users need to know what methods are available within `MooseApp` (not its parameters, but it's c++ methods). If we require our target demographic to understand deeply what `MooseApp` is, we have failed for a much larger reason.

I'm trying to help design-level framework developers here. Which is (hopefully) a very small subset of our user base. These are the people that API-level restrictions target. 

I don't see why we would not to benefit both groups.

Reply:
...and for a little more detail: I've added keys in these two specific places because I have exact thoughts in mind about how a specific individual or two would take advantage of an API in a way that would greatly frustrate me...

Reply:
Yes, this is a link between two classes, `MeshGeneatorMesh` and `MooseApp` that novice MOOSE developers should not be touching. And it's even more specific than that ... it's a link for a single API call whose location (hopefully) may never change. And if it does change it should only be done by a core MOOSE developer. And if they don't understand a passkey idiom for limiting data access, then they shouldn't be a core MOOSE developer

Reply:
It's a hacky idiom and it's not really needed. Again, MooseApp can know which stage it's currently in so it can restrict API use that way.

Also this class being public, cant anyone get the key by just building a dummy `MeshGeneratorMesh` ?

I ll let you ponder that comment on who should be a core developer

Reply:
> Also this class being public, cant anyone get the key by just building a dummy MeshGeneratorMesh ?

This API abuse is easier to see as a reviewer that may not have as much context to this system

Reply:
> I ll let you ponder that comment on who should be a core developer

I've pondered it and I feel good about it. Maybe someone doesn't understand passkey right off the bat, but if they are curious they can figure it out in under a half hour. I hope a core MOOSE developer is curious.

Reply:
@GiudGiud there are two in favor of the current design and you against. If you insist on a change I would propose that you submit a pull request against this PR to what you would like this to look like instead and we can review what you propose. Then we can weigh the two options

Reply:
>  I hope a core MOOSE developer is curious.

ok I m fine with that. 

Additionally, for setting parent/child MGs, most apps inherit MooseApp, so they can be cast as their parent class, be a friend and create the key? No need to even create a dummy MooseApp 

This solution ends up relying on code complexity to lock the API but we could achieve stricter control without it, with the traditional asserts of mooseApp->whatAmIDoingRightNow == something_that_makes_sense_for_this

So in summary:
- locking the final mesh generator api is not really that important (who abuses that, and what does that even give them?)
- locking the parent/child meshgenerator status is only a tad more important. Dependency resolution will still work if we add more parents or children, and dependency resolution fails gracefully anyway.

and both locks are easily circumvented.

Reply:
> Additionally, for setting parent/child MGs, most apps inherit MooseApp, so they can be cast as their parent class, be a friend and create the key? No need to even create a dummy MooseApp

Which, again, is a glaringly obvious concern for a reviewer that has no context of the MG system.

> This solution ends up relying on code complexity to lock the API but we could achieve stricter control without it, with the traditional asserts of mooseApp->whatAmIDoingRightNow == something_that_makes_sense_for_this

I'd be glad to have these too, but I've spent enough time trying to reach a middle ground here that I have no more gas in the tank to add more. 

> So in summary:

As I said above, I picked these methods in particular to protect as their importance to dependency resolution could cause someone to easily add in generators that do not prescribe to the basic assumptions I make about dependency ordering. I disagree with the "importance" here. We've all made bad design decisions that we could not have anticipated that have bitten us in the butt in the future. My priority right now is trying to lock down the assumptions I have made in this design.

> and both locks are easily circumvented.

Just like I could hack around the current task in `MooseApp`. Most everything is circumventable if you're okay with undefined behavior.

Reply:
Ok here s my proposal for your consideration
https://github.com/loganharbour/moose/pull/6

Reply:
Like I said above:

> I'd be glad to have these too

The problem is that this still allows a generator to add its own dependencies which significantly ruins some assumptions I have made about sub generators.

I'm happy to have both.

Reply:
Specifically for that, we could make the two addChild/Parent routines private and make them friend of MooseApp ?

This should ban this scenario.
And the hack around it is similar complexity as hacking the key with a MooseApp.

Reply:
ok it's not addChild/Parent that should be friend, it's MeshGeneratorSystem(previously MooseApp)::createMeshGenerator that should be a friend of MeshGenerator

Reply:
You can't make only a single method friendable. That's the point of the key. I don't want to give MooseApp access to every protected/private method/variable in MeshGenerator. That's much much worse.

Reply:
I tried the above unfortunately. You can’t declare private methods as friends.

Reply:
In summary, both methodologies work in this current case. Given that, I might slightly prefer @GiudGiud's suggestion given that it doesn't require a new idiom...but again both are equally completely safe in this case.

Reply:
I really like @dschwen’s idea of https://github.com/idaholab/moose/pull/23353#discussion_r1120331170

That is, OnlyAccessibleBy<MeshGeneratorSystem> in this case. I think it reads well

Reply:
> Consequently, contrary to what's been said previously, there is currently no way ... can muck things up.

I think you just need a pointer to a MeshGeneratorSystem to create the key (which everyone has through the app?). So there's a way to muck things up. But as mentioned the "friend routine" design isnt perfect either. 

I think after today's meeting the point is moot. I should have argued less about this and got more reviewing done, we d be finished by now. 

Let s keep working to get this through and have intelligently constructed, optimally memory-managed mesh generators and of course, working mesh metadata passing, the original intent of this

Reply:
`OnlyAccessibleBy<MeshGeneratorSystem>`
where are you going to define this?

APILockingInterface.C ? (better name most likely)

Reply:
> I think you just need a pointer to a MeshGeneratorSystem to create the key (which everyone has through the app?).

This isn't accurate. Only code within `MeshGeneratorSystem` methods can create the key

Reply:
> I tried the above unfortunately. You can’t declare private methods as friends.

Oh so my "paranoid" suggestion as well as @GiudGiud's suggestion of just friending the method simply won't compile. In that case yea I think I'd prefer the pass-key idiom which at least only exposes a couple methods for possible danger (if they were used in `MeshGeneratorSystem` public/protected methods) as opposed to all the `MeshGenerator` data/methods (through non-key-based `friend class MeshGeneratorSystem` in `MeshGenerator`).

Reply:
> This isn't accurate. Only code within MeshGeneratorSystem methods can create the key

The key is a public attribute within that class, I m not sure why that would be. I guess I have not tried.

anyhow we re not doing a good job at moving on boys

Reply:
Default access settings in classes are `private`. So `AddParentChildKey::AddParentChildKey` (both default and copy) constructors are `private`. E.g. no one can construct `AddParentChildKey` ... except the `friend class MeshGeneratorSystem`, e.g. `MeshGeneratorSystem` methods.

I just want us all to understand the C++ at play here 😄 Especially if we are going to be using the pass-key idiom